### ISSUE TITLE
Consider updating `release.json` format to add another level of indexing

### ISSUE THREAD

**Issue by richlander** [dotnet/core#9205]
The `release.json` files grow without bound (other than EOL). That's not a great architecture.

This file: https://github.com/dotnet/core/blob/main/release-notes/9.0/releases.json

Could look more like: https://github.com/dotnet/core/blob/main/release-notes/releases-index.json

And then have a `release.json` file in a directory like: https://github.com/dotnet/core/tree/main/release-notes/8.0/8.0.1

For most use cases, this approach would dramatically reduce the number of bytes downloaded, at least by a naive JSON reader like a serializer. It just occurred to me now, but much of the (example) premise of my [Convenience of System.Text.Json](https://devblogs.microsoft.com/dotnet/the-convenience-of-system-text-json/) post could be underminded by doing this.

**Comment by mairaw** [comment]
This sort of change would probably warrant a breaking-change announcement given that other people rely on that file for many things I'm not even aware of. Does Azure for example consume that sort of information?

The website is a given but I'm sure there are others.

**Comment by richlander** [comment]
For sure. We can do better. Also, if we were to come up with a short-link scheme, now would be the time to merge that in.

**Comment by Falco20019** [comment]
@richlander I did it testwise for .NET 8: https://github.com/Falco20019/core/commit/c80f4095458328d1502549c8d4150342aeccd8c2

.NET 9 already started to add the `preview1` folder below `preview` while .NET 8 and prior only had all markdown files in there directly and a named subfolder in `preview/api-diff`. I therefore would propose to clean this up as well. This also means that the `release-notes` for previews will change, but since they are linked in the updated `releases.json`, this should be fine.

My proposal would only leave `release-date`, `release-version`, `security` and `release-notes` in the `releases.json` of the channel and move all other (and duplicate) everything that was previously an element of the `releases` array. I added a `release.json` (attention: singular) short-link to those elements instead.

**Comment by richlander** [comment]
One way to avoid the breaking change is to create a peer to `releases-index.json`, with two options:

- Stop adding new releases to file at .NET 8 (remove .NET 9 from it).
- Continue publishing information to both flavors forever.

I don't think picking a date where we break the format is a good idea. Those plans never work.

**Comment by leecow** [comment]
Interesting ideas. A per-patch release file (e.g. /8.0.1/release.json) feels unappealing though that may be because I'm not understanding something. 

@richlander, can you expand on what you mean by a 'peer' of releases-index.json?

To avoid breaking existing users (and there's **a lot** of them), perhaps we could create the `release.json` (latest patch version only) alongside the `releases.json` (cumulative, maintaining the major release history).

**Comment by richlander** [comment]
Backing up, the large files we produce seem like an anti-pattern to me, in two ways:

- Almost no one ever needs all (or most) of the data in our `releas.json` files.
- Large JSON files should generally be avoided

As I said, I wrote a whole blog post on making reading these files efficient. It took a lot of effort. We shouldn't require our users to do that.

**Comment by leecow** [comment]
I don't disagree with any of that, though we have used the historical information in the 'complete' releases.json on a number of occasions. The clarifying question is along the lines of mitigating the breaking change while making this better for consumers. Something like:

- Implement a single, channel-located (e.g. https://github.com/dotnet/core/blob/main/release-notes/9.0/release.json) containing only the latest channel release information. Same schema as releases.json but only the latest release data.
- releases.json remains as-is but all documentation is updated to direct folks to the single file.
- Decide if and when the releases-index.json should be updated to point to the singleton providing a functional migration point to the new data for consumers that rely on the index for discovery.

**Comment by richlander** [comment]
> we have used the historical information in the 'complete' releases.json on a number of occasions

I don't see why this would ever be a problem. If there was a 9.0 index, you'd be able to get all the same information.

**Comment by richlander** [comment]
This probably isn't that hard. We just need to create another link here:

https://github.com/dotnet/core/blob/f424bc4f5ddf45ceefe1c8a538e29ff62c6c8d4c/release-notes/releases-index.json#L14

Like:

```json
        {
            "channel-version": "9.0",
            "latest-release": "9.0.0-preview.1",
            "latest-release-date": "2024-02-13",
            "security": false,
            "latest-runtime": "9.0.0-preview.1.24080.9",
            "latest-sdk": "9.0.100-preview.1.24101.2",
            "product": ".NET",
            "release-type" : "sts",
            "support-phase": "preview",
            "latest-release": "https://dotnetcli.blob.core.windows.net/dotnet/release-metadata/9.0/latest-release.json",
            "releases.json": "https://dotnetcli.blob.core.windows.net/dotnet/release-metadata/9.0/releases.json"
        }
```

**Comment by leecow** [comment]
Precisely what I was suggesting, thanks.

**Comment by Falco20019** [comment]
We can also add another `releases-index.json` per channel and link them in the main index. Than it's up to the caller if they follow the `releases.json` or `releases-index.json` link. No breaks involved.

**Comment by richlander** [comment]
I was also thinking about that. I think that's a good idea. We can then position the big `releases.json` as a legacy option.

**Comment by Falco20019** [comment]
One example I prepared a couple days ago:

https://github.com/Falco20019/core/commit/d6bb6275cd6170e89369dc3d01b3b49f5b555354

Just think of the `9.0/releases.json` to be named `9.0/releases-index.json` instead and the `9.0/releases.json` to look unchanged. Will adjust it tomorrow since I'm out of office today.

**Comment by richlander** [comment]
> Just think of the `9.0/releases.json` to be named `9.0/releases-index.json` instead and the `9.0/releases.json` to look unchanged.

Do you have a typo there? You mention `9.0/releases.json` twice.

We'd be very hesitant to change the 8.0 preview content. It is now in the past and putting effort into that or breaking people isn't desirable. We want to focus our efforts on forward-looking improvements.

**Comment by Falco20019** [comment]
No typo, maybe just written confusing. This was in reference to the linked commit. The change listed says `9.0/releases.json` but should be `9.0/releases-index.json`. I prepared it before that was discussed.

And the second mention meant that beside the name, the original one is kept as discussed here.

**Comment by richlander** [comment]
I understand your wording now.

My thinking is that we'd start this scheme with (for example) 8.0.3 and just move forward. Backporting the scheme would be nice, but it isn't clear that it is needed.

**Comment by Falco20019** [comment]
I can have a look since I already did it test wise for the full .NET 8 and 9 releases. I think having .NET 6 through 9 should be enough, but as it just feels incomplete, I might try to generate the rest as well for completion sake.

Sorry if my wording is sometimes confusing. I’m no native speaker.

**Comment by richlander** [comment]
> Sorry if my wording is sometimes confusing. I’m no native speaker.

Hearing you loud and clear.

Let's validate we have a good model for the in-support versions and then if you want to backfill date, you can.

**Comment by martincostello** [comment]
Throwing my hat into the ring as a consumer of this data.

I maintain a GitHub Action that consumes this data for automating patches: [martincostello/update-dotnet-sdk](https://github.com/martincostello/update-dotnet-sdk).

I have various other things built on top of this data for automation purposes as well (_"Is your internal application running the latest patch version of .NET?"_, _"Download the latest two versions of the .NET SDK"_, _"Is this version of .NET still in support?"_).

If things are going to be changed in a breaking manner, then for myself at least, I'd need a few weeks' notice of the final format (and ideally a side-by-side deployment period) to update all these things to the new data format.

**Comment by Falco20019** [comment]
@martincostello We will make sure to only introduce the new format in a non-breaking manner. Any existing file must not be changed in a way not covered/permitted by JSON schemas. Adding fields might require an updated JSON schema, but should not break anyone.

Most changes will be in new files and should be introduced in parallel (maybe even marked as experimental for now) to gather input.

**Comment by leecow** [comment]
Another point to mention is that the existing assets are machine generated and maintained, and the new ones will be as well. While we can iterate over ideas in 'experimental' PRs, the final PR implementing what we decide upon will be the result of tooling updates.

**Comment by martincostello** [comment]
Being able to determine which repo(s) a CVE applies to would definitely have value. For example, a CVE related to WPF wouldn't be relevant for an ASP.NET Core app.

**Comment by richlander** [comment]
Now that we have fancy markdown releases notes in this repo, I'm wondering if we should move our CVEs from dotnet/announcements to a `cves` directory (peer to `9.0`), with years as children. We'd then rely on our digest posts to document the CVEs as a group. 

We document CVEs as markdown, but don't share it. That's not intentional, but a function of the way GitHub issues works. We publish markdown and you get HTML. A major goal of this new effort is to share markdown since it is a lot more flexible for re-use than HTML.

We're currently working on a `cve.json` format as part of this project. Each CVE object would then link to the markdown file (with a regular GH URL, but switching that to the `raw` variant should be trivial should that be desired).

Good idea?

I also took the opportunity to clarify the license of all of this content. I hope that's favorable.

- Markdown release notes: https://github.com/dotnet/core/tree/main/release-notes/9.0/preview/preview3
- CVE post: https://github.com/dotnet/announcements/issues/303
- Digest post: https://github.com/dotnet/announcements/issues/302
- License: https://github.com/dotnet/core/blob/main/release-notes/license-information.md

**Comment by richlander** [comment]
This issue is too long and some of the conversation (my fault) is confusing. I'm going to create more narrow issues going forward on this general topic.
